### PR TITLE
[InstSimplify] fold uno/ord comparison if fpclass is always NaN

### PR DESCRIPTION
**Pull request by AlexMaclean**
In InstSimplify we already fold `fcmp ord/uno` to a constant when both operands are known to be non-NaN. This change slightly generalizes this to also handle the case where either of the operands is known to always be NaN. 

Proof: https://alive2.llvm.org/ce/z/AhCmJN

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8111]: It should be folded into a constant boolean value.
```suggestion
        return replaceInstUsesWith(I, ConstantInt::getBool(Pred == CmpInst::FCMP_UNO));
```

**Review by dtcxzyw — COMMENTED** [review]
Please attach alive2 proofs and pre-commit tests.

**Comment by AlexMaclean** [comment]
> Please attach alive2 proofs and pre-commit tests.

Okay, I've attached a proof, could you explain further about what you mean by pre-commit tests?

**Comment by nikic** [comment]
> > Please attach alive2 proofs and pre-commit tests.
> 
> Okay, I've attached a proof, could you explain further about what you mean by pre-commit tests?

See https://llvm.org/docs/InstCombineContributorGuide.html#precommit-tests.

**Comment by AlexMaclean** [comment]
> > > Please attach alive2 proofs and pre-commit tests.
> > 
> > 
> > Okay, I've attached a proof, could you explain further about what you mean by pre-commit tests?
> 
> See https://llvm.org/docs/InstCombineContributorGuide.html#precommit-tests.

Thanks, I've split the tests out into a separate commit as described, please let me know if I've made any mistake.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8110]: It would be better to move this fold into InstSimplify:
https://github.com/llvm/llvm-project/blob/0b9f2847da79298ed09c29493245113f02b32d9f/llvm/lib/Analysis/InstructionSimplify.cpp#L4115-L4123

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:8112]: Don't use `Builder.getInt1`. It will fail when folding a vector fcmp.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

**Comment by dtcxzyw** [comment]
Please update the title and description, since this PR folds the pattern into a constant instead of canonicalizing its operands.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (AlexMaclean) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Moved.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp
Comment (AlexMaclean) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Fixed.

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (dtcxzyw) [inline llvm/lib/Analysis/InstructionSimplify.cpp:4120]:
```suggestion
    KnownFPClass RHSClass = computeKnownFPClass(RHS, fcAllFlags, /*Depth=*/0, Q);
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers :)

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**
